File: execution/fetch_github_repos.js
const fs = require('fs');
const path = require('path');
const https = require('https');

// Helper to fetch JSON from API using native https module
function fetchJson(url, token) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const options = {
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      method: 'GET',
      headers: {
        'User-Agent': 'NodeJS-Agent',
        'Accept': 'application/vnd.github.v3+json'
      }
    };

    if (token) {
      options.headers['Authorization'] = `token ${token}`;
    }

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200) {
          try {
            resolve(JSON.parse(data));
          } catch (e) {
            reject(new Error(`Failed to parse response: ${e.message}`));
          }
        } else {
          reject(new Error(`API Error: ${res.statusCode} - ${res.statusMessage}`));
        }
      });
    });

    req.on('error', (err) => {
      reject(err);
    });

    req.end();
  });
}

async function main() {
  // Simple .env parser
  const envVars = {};
  const envPath = path.join(__dirname, '..', '.env');
  if (fs.existsSync(envPath)) {
    const content = fs.readFileSync(envPath, 'utf-8');
    content.split('\n').forEach(line => {
      line = line.trim();
      if (line && !line.startsWith('#')) {
        const parts = line.split('=');
        if (parts.length >= 2) {
          const key = parts[0].trim();
          const val = parts.slice(1).join('=').trim();
          envVars[key] = val;
        }
      }
    });
  }

  // Inputs
  const username = envVars['GITHUB_USERNAME'] || 'google';
  const token = envVars['GITHUB_TOKEN'] || process.env.GITHUB_TOKEN;
  const maxRepos = parseInt(envVars['MAX_REPOS'] || '5', 10);
  const outputPath = envVars['OUTPUT_PATH'] || 'github_report.md';

  console.log(`Starting GitHub analysis for user: '${username}'...`);

  try {
    // 1. Fetch User Info
    const userUrl = `https://api.github.com/users/${username}`;
    const userData = await fetchJson(userUrl, token);

    // 2. Fetch Repositories
    const reposUrl = `https://api.github.com/users/${username}/repos?per_page=100`;
    const reposData = await fetchJson(reposUrl, token);

    // Sort by stars descending
    const sortedRepos = reposData.sort((a, b) => (b.stargazers_count || 0) - (a.stargazers_count || 0));
    const topRepos = sortedRepos.slice(0, maxRepos);

    // Stats
    const totalStars = reposData.reduce((acc, repo) => acc + (repo.stargazers_count || 0), 0);
    const languages = {};
    reposData.forEach(repo => {
      const lang = repo.language;
      if (lang) {
        languages[lang] = (languages[lang] || 0) + 1;
      }
    });

    const sortedLanguages = Object.entries(languages).sort((a, b) => b[1] - a[1]);

    // 3. Generate Report Content
    const nowStr = new Date().toISOString().replace('T', ' ').substring(0, 19);
    let reportContent = `# GitHub Profile Analysis: ${userData.name || username}

Generated on: ${nowStr}

## Profile Details
- **Username**: ${userData.login}
- **Bio**: ${userData.bio || 'N/A'}
- **Location**: ${userData.location || 'N/A'}
- **Public Repositories**: ${userData.public_repos}
- **Followers**: ${userData.followers}
- **Total Stars Accumulated**: ${totalStars}

## Top Languages
${sortedLanguages.slice(0, 3).map(([lang, count]) => `**${lang}** (${count})`).join(', ')}

## Top ${topRepos.length} Repositories (by Stars)

| Repository | Stars | Forks | Language | Description |
|---|---|---|---|---|
`;

    topRepos.forEach(repo => {
      const name = repo.name;
      const htmlUrl = repo.html_url;
      const stars = repo.stargazers_count;
      const forks = repo.forks_count;
      const lang = repo.language || 'N/A';
      const desc = repo.description || 'No description provided.';
      reportContent += `| [${name}](${htmlUrl}) | ${stars} | ${forks} | ${lang} | ${desc} |\n`;
    });

    const absoluteOutputPath = path.isAbsolute(outputPath) ? outputPath : path.join(__dirname, '..', outputPath);
    fs.writeFileSync(absoluteOutputPath, reportContent, 'utf-8');
    console.log(`Analysis completed successfully! Report saved to ${outputPath}`);

  } catch (error) {
    console.error(`Error executing script: ${error.message}`);
    process.exit(1);
  }
}

main();

File: execution/fetch_github_repos.py
import os
import json
import urllib.request
import urllib.error
import sys
from datetime import datetime

def fetch_data(url, token=None):
    """Fetches JSON data from the GitHub API using urllib."""
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
        "Accept": "application/vnd.github.v3+json"
    }
    if token:
        headers["Authorization"] = f"token {token}"
        
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req) as response:
            if response.status == 200:
                return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        print(f"Error fetching data from API: {e.code} - {e.reason}", file=sys.stderr)
        if e.code == 403:
            print("API Rate limit exceeded or unauthorized. Please set GITHUB_TOKEN in .env.", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"Connection error: {e.reason}", file=sys.stderr)
        sys.exit(1)

def main():
    # Load configuration from environment or command line
    # (Simple .env parser to avoid external dependencies like python-dotenv)
    env_vars = {}
    if os.path.exists(".env"):
        with open(".env", "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#"):
                    parts = line.split("=", 1)
                    if len(parts) == 2:
                        env_vars[parts[0].strip()] = parts[1].strip()

    # Inputs (can be passed via env variables, config, or arguments)
    username = env_vars.get("GITHUB_USERNAME", "google")
    token = env_vars.get("GITHUB_TOKEN", os.getenv("GITHUB_TOKEN"))
    max_repos = int(env_vars.get("MAX_REPOS", 5))
    output_path = env_vars.get("OUTPUT_PATH", "github_report.md")

    print(f"Starting GitHub analysis for user: '{username}'...")
    
    # 1. Fetch User Info
    user_url = f"https://api.github.com/users/{username}"
    user_data = fetch_data(user_url, token)
    
    # 2. Fetch User Repositories
    repos_url = f"https://api.github.com/users/{username}/repos?per_page=100"
    repos_data = fetch_data(repos_url, token)
    
    # Sort repos by stargazers_count descending
    sorted_repos = sorted(repos_data, key=lambda x: x.get("stargazers_count", 0), reverse=True)
    top_repos = sorted_repos[:max_repos]
    
    # Calculate stats
    total_stars = sum(repo.get("stargazers_count", 0) for repo in repos_data)
    languages = {}
    for repo in repos_data:
        lang = repo.get("language")
        if lang:
            languages[lang] = languages.get(lang, 0) + 1
    
    sorted_languages = sorted(languages.items(), key=lambda x: x[1], reverse=True)
    
    # 3. Generate Markdown Report
    now_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    report_content = f"""# GitHub Profile Analysis: {user_data.get('name') or username}

Generated on: {now_str}

## Profile Details
- **Username**: {user_data.get('login')}
- **Bio**: {user_data.get('bio') or 'N/A'}
- **Location**: {user_data.get('location') or 'N/A'}
- **Public Repositories**: {user_data.get('public_repos')}
- **Followers**: {user_data.get('followers')}
- **Total Stars Accumulated**: {total_stars}

## Top Languages
{', '.join([f"**{lang}** ({count})" for lang, count in sorted_languages[:3]])}

## Top {len(top_repos)} Repositories (by Stars)

| Repository | Stars | Forks | Language | Description |
|---|---|---|---|---|
"""
    
    for repo in top_repos:
        name = repo.get("name")
        html_url = repo.get("html_url")
        stars = repo.get("stargazers_count")
        forks = repo.get("forks_count")
        lang = repo.get("language") or "N/A"
        desc = repo.get("description") or "No description provided."
        
        report_content += f"| [{name}]({html_url}) | {stars} | {forks} | {lang} | {desc} |\n"
        
    # Write to output file
    try:
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(report_content)
        print(f"Analysis completed successfully! Report saved to {output_path}")
    except Exception as e:
        print(f"Error writing report: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()
